libfreerdp/core/peer.c

```diff
@@ -893,6 +893,225 @@ static BOOL freerdp_peer_send_channel_data(freerdp_peer* client, UINT16 channelI
+	UINT32 passwordLength;
+	UINT32 targetFQDNLength;
+	UINT32 targetNetBiosNameLength;
+	UINT32 targetNetAddressesLength;
+	UINT32* targetNetAddressesWLength;
+
+	LPWSTR targetNetAddressW = NULL;
+	LPWSTR userNameW = NULL;
+	LPWSTR domainW = NULL;
+	LPWSTR passwordW = NULL;
+	LPWSTR targetFQDNW = NULL;
+	LPWSTR targetNetBiosNameW = NULL;
+	LPWSTR* targetNetAddressesW = NULL;
+
+	length = 12; /* Flags (2) + length (2) + sessionId (4)  + redirection flags (4) */
+	redirFlags = 0;
+
+	if(targetNetAddress)
+	{
+		redirFlags |= LB_TARGET_NET_ADDRESS;
+
+		ConvertToUnicode(CP_UTF8, 0, (LPCSTR)targetNetAddress, -1, &targetNetAddressW, 0);
+		targetNetAddressLength = (strlen(targetNetAddress) + 1) * sizeof(WCHAR);
+
+		length += 4 + targetNetAddressLength;
+	}
+
+	if(routingToken)
+	{
+		redirFlags |= LB_LOAD_BALANCE_INFO;
+		loadBalanceInfoLength = 13 + strlen(routingToken) + 2; /* Add routing token prefix and suffix */
+		length += 4 + loadBalanceInfoLength;
+	}
+
+	if(userName)
+	{
+		redirFlags |= LB_USERNAME;
+
+		ConvertToUnicode(CP_UTF8, 0, (LPCSTR)userName, -1, &userNameW, 0);
+		userNameLength = (strlen(userName) + 1) * sizeof(WCHAR);
+
+		length += 4 + userNameLength;
+	}
+
+	if(domain)
+	{
+		redirFlags |= LB_DOMAIN;
+
+		ConvertToUnicode(CP_UTF8, 0, (LPCSTR)domain, -1, &domainW, 0);
+		domainLength = (strlen(domain) + 1) * sizeof(WCHAR);
+
+		length += 4 + domainLength;
+	}
+
+	if(password)
+	{
+		redirFlags |= LB_PASSWORD;
+
+		ConvertToUnicode(CP_UTF8, 0, (LPCSTR)password, -1, &passwordW, 0);
+		passwordLength = (strlen(password) + 1) * sizeof(WCHAR);
+
+		length += 4 + passwordLength;
+	}
+
+	if(targetFQDN)
+	{
+		redirFlags |= LB_TARGET_FQDN;
+
+		ConvertToUnicode(CP_UTF8, 0, (LPCSTR)targetFQDN, -1, &targetFQDNW, 0);
+		targetFQDNLength = (strlen(targetFQDN) + 1) * sizeof(WCHAR);
+
+		length += 4 + targetFQDNLength;
+	}
+
+	if(targetNetBiosName)
+	{
+		redirFlags |= LB_TARGET_NETBIOS_NAME;
+
+		ConvertToUnicode(CP_UTF8, 0, (LPCSTR)targetNetBiosName, -1, &targetNetBiosNameW, 0);
+		targetNetBiosNameLength = (strlen(targetNetBiosName) + 1) * sizeof(WCHAR);
+
+		length += 4 + targetNetBiosNameLength;
+	}
+
+	if(tsvUrl)
+	{
+		redirFlags |= LB_CLIENT_TSV_URL;
+		length += 4 + tsvUrlLength;
+	}
+
+	if(targetNetAddresses)
+	{
+		int i;
+
+		redirFlags |= LB_TARGET_NET_ADDRESSES;
+
+		targetNetAddressesLength = 0;
+		targetNetAddressesW = calloc(targetNetAddressesCount, sizeof(LPWSTR));
+		targetNetAddressesWLength = calloc(targetNetAddressesCount, sizeof(UINT32));
+		for (i = 0; i < targetNetAddressesCount; i++)
+		{
+			ConvertToUnicode(CP_UTF8, 0, (LPCSTR)targetNetAddresses[i], -1, &targetNetAddressesW[i], 0);
+			targetNetAddressesWLength[i] = (strlen(targetNetAddresses[i]) + 1) * sizeof(WCHAR);
+			targetNetAddressesLength += 4 + targetNetAddressesWLength[i];
+		}
+
+		length += 4 + 4 + targetNetAddressesLength;
+	}
+
+	Stream_Write_UINT16(s, 0);
+	Stream_Write_UINT16(s, SEC_REDIRECTION_PKT);
+	Stream_Write_UINT16(s, length);
```

Comment:
We shall have a `if (!Stream_EnsureRemainingCapacity(s, length)) ...` here to ensure that the stream has enough room for what we want to write in it